server/sync/monitor/monitor.go

```diff
@@ -0,0 +1,348 @@
+// Copyright AGNTCY Contributors (https://github.com/agntcy)
+// SPDX-License-Identifier: Apache-2.0
+
+package monitor
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	corev1 "github.com/agntcy/dir/api/core/v1"
+	"github.com/agntcy/dir/server/store/oci"
+	ociconfig "github.com/agntcy/dir/server/store/oci/config"
+	"github.com/agntcy/dir/server/sync/monitor/config"
+	"github.com/agntcy/dir/server/types"
+	"github.com/agntcy/dir/server/types/adapters"
+	"github.com/agntcy/dir/utils/logging"
+	"oras.land/oras-go/v2/registry/remote"
+)
+
+var logger = logging.Logger("sync/monitor")
+
+// MonitorService manages registry monitoring based on active sync operations.
+//
+//nolint:revive
+type MonitorService struct {
+	// Configuration
+	db            types.DatabaseAPI
+	store         types.StoreAPI
+	ociConfig     ociconfig.Config
+	checkInterval time.Duration
+
+	// Monitoring state
+	mu            sync.RWMutex
+	isRunning     bool
+	lastSnapshot  *RegistrySnapshot
+	ticker        *time.Ticker
+	cancelMonitor context.CancelFunc
+
+	// Sync management
+	activeSyncs map[string]struct{} // Track active sync operations
+
+	// ORAS repository client
+	repo *remote.Repository
+}
+
+// NewMonitorService creates a new monitor service.
+func NewMonitorService(db types.DatabaseAPI, store types.StoreAPI, ociConfig ociconfig.Config, monitorConfig config.Config) (*MonitorService, error) {
+	// Create ORAS repository client
+	repo, err := oci.NewORASRepository(ociConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ORAS repository client: %w", err)
+	}
+
+	return &MonitorService{
+		db:            db,
+		store:         store,
+		ociConfig:     ociConfig,
+		checkInterval: monitorConfig.CheckInterval,
+		activeSyncs:   make(map[string]struct{}),
+		repo:          repo,
+	}, nil
+}
+
+// Stop gracefully shuts down the monitor service.
+func (s *MonitorService) Stop() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	logger.Info("Stopping monitor service")
+
+	// Stop monitoring if active
+	if s.isRunning {
+		logger.Info("Stopping registry monitoring")
+
+		// Cancel monitoring
+		if s.cancelMonitor != nil {
+			s.cancelMonitor()
+		}
+
+		s.ticker.Stop()
+
+		// Update state
+		s.isRunning = false
+	}
+
+	// Clear active syncs
+	s.activeSyncs = make(map[string]struct{})
+
+	logger.Info("Monitor service stopped")
+
+	return nil
+}
+
+// StartSyncMonitoring begins monitoring when a sync operation starts.
+func (s *MonitorService) StartSyncMonitoring(syncID string) error {
```

Comment:
can we pass the context here in any way, otherwise this may hang forever unless on force-exit. we can pass the actual server context to the MonitorService and reuse it here as well